### PR TITLE
address page: Add charts

### PR DESCRIPTION
**Pull request by dmigwi**
This PR adds a Histogram and a Chart showing Amount flow.

**View 1 (Main addresses page view)**
(Main addresses page view)

**View 2 (Tx Types Distribution Chart)**
(Tx Types Distribution Chart)

**View 3 (Amount Flow Chart)**
screen shot 2018-08-08 at 05 25 13

### CODE REVIEW

**Comment by chappjc** [comment]
Slated for v3.1

**Comment by dmigwi** [comment]
I have successfully tested for the upgrade needed in this PR.
```
2018-08-12 14:21:49.133 [INF] PSQL: Adding column tx_type to table vins...
2018-08-12 14:21:49.161 [INF] PSQL: found the best block at height: 263820
2018-08-12 14:21:49.162 [INF] PSQL:  - Retrieving data from txs table between height 1 and 50000 ...
2018-08-12 14:21:54.315 [INF] PSQL: Populating vins table with data between height 1 and 50000 ...
2018-08-12 14:35:16.646 [INF] PSQL:  - Retrieving data from txs table between height 50001 and 100000 ...
2018-08-12 14:35:21.691 [INF] PSQL: Populating vins table with data between height 50001 and 100000 ...
2018-08-12 14:54:59.600 [INF] PSQL:  - Retrieving data from txs table between height 100001 and 150000 ...
2018-08-12 14:55:05.169 [INF] PSQL: Populating vins table with data between height 100001 and 150000 ...
2018-08-12 15:21:08.694 [INF] PSQL:  - Retrieving data from txs table between height 150001 and 200000 ...
2018-08-12 15:21:16.159 [INF] PSQL: Populating vins table with data between height 150001 and 200000 ...
2018-08-12 15:39:09.317 [INF] PSQL:  - Retrieving data from txs table between height 200001 and 250000 ...
2018-08-12 15:39:15.575 [INF] PSQL: Populating vins table with data between height 200001 and 250000 ...
2018-08-12 15:55:08.825 [INF] PSQL:  - Retrieving data from txs table between height 250001 and 263820 ...
2018-08-12 15:55:10.567 [INF] PSQL: Populating vins table with data between height 250001 and 263820 ...
2018-08-12 15:59:19.245 [INF] PSQL:  - 11797237 rows in vins table (tx-type addition/histogram upgrade) were successfully upgraded in 5849.75 sec
2018-08-12 15:59:19.291 [INF] PSQL: Adding column tx_type to table addresses...
2018-08-12 15:59:19.325 [INF] PSQL: found the best block at height: 263820
2018-08-12 15:59:19.325 [INF] PSQL:  - Retrieving data from txs table between height 1 and 50000 ...
2018-08-12 15:59:22.403 [INF] PSQL: Populating addresses table with data between height 1 and 50000 ...
2018-08-12 16:35:11.378 [INF] PSQL:  - Retrieving data from txs table between height 50001 and 100000 ...
2018-08-12 16:35:16.948 [INF] PSQL: Populating addresses table with data between height 50001 and 100000 ...
2018-08-12 17:17:58.972 [INF] PSQL:  - Retrieving data from txs table between height 100001 and 150000 ...
2018-08-12 17:18:05.326 [INF] PSQL: Populating addresses table with data between height 100001 and 150000 ...
2018-08-12 18:08:53.805 [INF] PSQL:  - Retrieving data from txs table between height 150001 and 200000 ...
2018-08-12 18:08:59.356 [INF] PSQL: Populating addresses table with data between height 150001 and 200000 ...
2018-08-12 19:11:32.585 [INF] PSQL:  - Retrieving data from txs table between height 200001 and 250000 ...
2018-08-12 19:11:39.424 [INF] PSQL: Populating addresses table with data between height 200001 and 250000 ...
2018-08-12 20:04:40.947 [INF] PSQL:  - Retrieving data from txs table between height 250001 and 263820 ...
2018-08-12 20:04:51.716 [INF] PSQL: Populating addresses table with data between height 250001 and 263820 ...
2018-08-12 20:19:22.458 [INF] PSQL:  - 22919723 rows in addresses table (tx-type addition/histogram upgrade) were successfully upgraded in 15602.24 sec
2018-08-12 20:19:22.487 [INF] PSQL: Modified the vouts table version to 3.5.0
2018-08-12 20:19:23.522 [INF] PSQL: Modified the addresses table version to 3.5.0
2018-08-12 20:19:23.524 [INF] PSQL: Modified the tickets table version to 3.5.0
2018-08-12 20:19:23.526 [INF] PSQL: Modified the agendas table version to 3.5.0
2018-08-12 20:19:23.527 [INF] PSQL: Modified the misses table version to 3.5.0
2018-08-12 20:19:23.528 [INF] PSQL: Modified the blocks table version to 3.5.0
2018-08-12 20:19:23.528 [INF] PSQL: Modified the transactions table version to 3.5.0
2018-08-12 20:19:24.603 [INF] PSQL: Modified the vins table version to 3.5.0
2018-08-12 20:19:24.605 [INF] PSQL: Modified the block_chain table version to 3.5.0
2018-08-12 20:19:24.606 [INF] PSQL: Modified the votes table version to 3.5.0
```
All that is pending is just modifying so charts and then all will be done.

**Comment by dmigwi** [comment]
The time is took to run this PR maybe longer than it will take for people using SSD hard disks.

**Comment by dmigwi** [comment]
Here are the final versions of the three main charts.

The main addresses page view remains as was in the above screenshots.

**Chart 1 (Transactions Types)**

screen shot 2018-08-17 at 04 59 26

**Chart 2 (Sent and Recieved)**

screen shot 2018-08-17 at 04 59 52


**Chart 3 (Unspent)**

screen shot 2018-08-17 at 05 00 18

**Comment by dmigwi** [comment]
@chappjc This PR feels ready for a full review and testing.

File: db/dbtypes/types.go [hunk 1]
```diff
@@ -331,6 +383,14 @@ type ChartsData struct {
 	Height     []uint64  `json:"height,omitempty"`
 	Pooled     []uint64  `json:"pooled,omitempty"`
 	Solo       []uint64  `json:"solo,omitempty"`
+	RegularTx  []uint64  `json:"regularTx,omitempty"`
+	Tickets    []uint64  `json:"tickets,omitempty"`
+	Votes      []uint64  `json:"votes,omitempty"`
+	RevokeTx   []uint64  `json:"revokeTx,omitempty"`
+	Amount     []float64 `json:"amount,omitempty"`
+	Recieved   []float64 `json:"received,omitempty"`
```
[the remark this answers is not in the document]
Comment (dmigwi) [reply db/dbtypes/types.go]: Ok. Noted

File: db/dcrpg/pgblockchain.go [hunk 2]
```diff
@@ -916,6 +916,31 @@ func (pgb *ChainDB) Store(blockData *blockdata.BlockData, msgBlock *wire.MsgBloc
...
+		return retrieveTxHistoryByType(pgb.db, address, timestamp)
+
+	case dbtypes.AmountFlow:
+		return retrieveTxHistoryByAmountFlow(pgb.db, address, timestamp)
+
+	case dbtypes.TotalUnspent:
+		return retrieveTxHistoryByUnspentAmount(pgb.db, address, timestamp)
+
+	default:
+		return nil, fmt.Errorf("Unknown error occured")
```
[the remark this answers is not in the document]
Comment (dmigwi) [reply db/dcrpg/pgblockchain.go]: ok. Noted.

File: db/dcrpg/queries.go [hunk 3]
```diff
@@ -1635,6 +1575,101 @@ func closeRows(rows *sql.Rows) {
...
+		items.Time = append(items.Time, blockTime)
+		items.RegularTx = append(items.RegularTx, regularTx)
+		items.Tickets = append(items.Tickets, tickets)
+		items.Votes = append(items.Votes, votes)
+		items.RevokeTx = append(items.RevokeTx, revokeTx)
+	}
+	return items, nil
+}
+
+// retrieveTxHistoryByAmount fetches the transaction amount flow i.e. recieved
```
[the remark this answers is not in the document]
Comment (dmigwi) [reply db/dcrpg/queries.go]: ok. Noted

File: db/dcrpg/upgrades.go [hunk 4]
```diff
@@ -61,33 +74,25 @@ func (pgb *ChainDB) CheckForAuxDBUpgrade(dcrdClient *rpcclient.Client) (bool, er
...
-		for it := range theseUpgrades {
-			upgradeSuccess, err := pgb.handleUpgrades(smartClient, theseUpgrades[it].upgradeType)
-			if err != nil || !upgradeSuccess {
-				return false, fmt.Errorf("failed to upgrade %s table to version %v. Error: %v",
-					theseUpgrades[it].TableName, toVersion, err)
-			}
-		}
-
-		// Bump version
-		if err := versionAllTables(pgb.db, toVersion); err != nil {
```
[the remark this answers is not in the document]
Comment (dmigwi) [reply db/dcrpg/upgrades.go]: ok. Makes sense

File: db/dcrpg/upgrades.go [hunk 5]
```diff
@@ -33,13 +33,26 @@ const (
...
 )
 
 type TableUpgradeType struct {
 	TableName   string
 	upgradeType tableUpgradeType
 }
 
+// histogramUpdate defines the fetched details from the transactions table that
+// are needed to undertake the histogram upgrade.
+type histogramUpdateType struct {
```
[the remark this answers is not in the document]
Comment (dmigwi) [reply db/dcrpg/upgrades.go]: ok. Noted

File: db/dcrpg/upgrades.go [hunk 6]
```diff
@@ -143,24 +133,34 @@ func (pgb *ChainDB) CheckForAuxDBUpgrade(dcrdClient *rpcclient.Client) (bool, er
...
-		// Bump version
-		if err := versionAllTables(pgb.db, toVersion); err != nil {
-			return false, fmt.Errorf("failed to bump version to %v: %v", toVersion, err)
+		// Go on to next upgrade
+		fallthrough
+
+	// Upgrade from 3.4.1 --> 3.5.0
+	case version.major == 3 && version.minor == 4 && version.patch == 1:
+		toVersion = TableVersion{3, 5, 0}
+		smartClient := rpcutils.NewBlockGate(dcrdClient, 10)
```
[the remark this answers is not in the document]
Comment (dmigwi) [reply db/dcrpg/upgrades.go]: Its needed to get the best block

[the remark this answers is not in the document]
Comment (chappjc) [reply db/dcrpg/upgrades.go]: I don't see the client used in either of these upgrades.  It comes from `(*ChainDB).HashDB`.

**Comment by chappjc** [comment]
g2g, just the conflict.

**Comment by dmigwi** [comment]
@chappjc I have fixed the merge conflicts

**Comment by chappjc** [comment]
Looks like the indexing of the address table is in the wrong spot:

```[INF] PSQL: Indexing on addresses table just to hasten the update
[INF] PSQL: Populating addresses table with data between height 1 and 50000 ...
[INF] PSQL:  - Retrieving data from txs table between height 50001 and 100000 ...
[INF] PSQL: Indexing on addresses table just to hasten the update
[WRN] PSQL: histogram upgrade on addresses table maybe
		slower since indexing failed: pq: relation "xxxxx_histogram" already exists
[INF] PSQL: Populating addresses table with data between height 50001 and 100000 ...
[INF] PSQL:  - Retrieving data from txs table between height 100001 and 150000 ...
[INF] PSQL: Indexing on addresses table just to hasten the update
[WRN] PSQL: histogram upgrade on addresses table maybe
		slower since indexing failed: pq: relation "xxxxx_histogram" already exists
[INF] PSQL: Populating addresses table with data between height 100001 and 150000 ...
[INF] PSQL:  - Retrieving data from txs table between height 150001 and 200000 ...
[INF] PSQL: Indexing on addresses table just to hasten the update
[WRN] PSQL: histogram upgrade on addresses table maybe
		slower since indexing failed: pq: relation "xxxxx_histogram" already exists
[INF] PSQL: Populating addresses table with data between height 150001 and 200000 ...
[INF] PSQL:  - Retrieving data from txs table between height 200001 and 250000 ...
[INF] PSQL: Indexing on addresses table just to hasten the update
[WRN] PSQL: histogram upgrade on addresses table maybe
		slower since indexing failed: pq: relation "xxxxx_histogram" already exists
[INF] PSQL: Populating addresses table with data between height 200001 and 250000 ...
[INF] PSQL:  - Retrieving data from txs table between height 250001 and 264201 ...
[INF] PSQL: Indexing on addresses table just to hasten the update
[WRN] PSQL: histogram upgrade on addresses table maybe
		slower since indexing failed: pq: relation "xxxxx_histogram" already exists
[INF] PSQL: Populating addresses table with data between height 250001 and 264201 ...
```

**Comment by dmigwi** [comment]
I have also made the address table indexing run outside the for loop.

**Comment by dmigwi** [comment]
I implemented the fix some minutes back but forgot to inform you https://github.com/decred/dcrdata/pull/597/commits/2e0490ee0886da23c1e3a85de77fe442e6fc42cf

**Pull request merged**